gabarit/template_nlp/nlp_project/package_name/models_training/model_aggregation.py

```diff
@@ -0,0 +1,440 @@
+        return self.list_classes[index_class]
+
+    def majority_vote(self, predictions: np.ndarray):
+        '''Gives the class corresponding to the most present prediction in the given predictions. 
+        In case of a tie, gives the first prediction (even if not the most present) 
+
+        Args:
+            predictions (np.ndarray) : The array containing the predictions of each model (shape (n_models)) 
+        Returns:
+            The prediction
+        '''
+        labels, counts = np.unique(predictions, return_counts=True)
+        votes = [(label, count) for label, count in zip(labels, counts)]
+        votes = sorted(votes, key=lambda x: x[1], reverse=True)
+        if len(votes) > 1 and votes[0][1] == votes[1][1]:
+            return predictions[0]
+        else:
+            return votes[0][0]
+
+    def all_predictions(self, predictions: np.ndarray) -> np.ndarray:
+        '''Calculates the sum of the arrays along axis 0 casts it to bool and then to int. 
+        Expects a numpy array containing only zeroes and ones. 
+        When used as an aggregation function, keeps all the prediction of each model (multi-labels)
+
+        Args:
+            predictions (np.ndarray) : Array of shape : (n_models, n_classes)
+        Return:
+            (np.ndarray) : The prediction
+        '''
+        return np.sum(predictions, axis=0, dtype=bool).astype(int)
+
+    def vote_labels(self, predictions: np.ndarray) -> np.ndarray:
+        '''Gives the result of majority_vote applied on the second axis. 
+        When used as an aggregation_function, for each class, performs a majority vote for the aggregated models. 
+        It gives a multi-labels result
+
+        Args:
+            (np.ndarray) : array of shape : (n_models, n_classes)
+        Return:
+            (np.ndarray) : prediction
+        '''
+        return np.apply_along_axis(self.majority_vote, 0, predictions)
+
+    def save(self, json_data: Union[dict, None] = None) -> None:
+        '''Saves the model
+
+        Kwargs:
+            json_data (dict): Additional configurations to be saved
+        '''
+        if json_data is None:
+            json_data = {}
+        # Save each trained and unsaved model
+        for i, model in enumerate(self.list_real_models):
+            if not self.list_models_trained[i] and model.trained:
+                model.save()
+
+        json_data['list_models'] = self.list_models.copy()
+        json_data['using_proba'] = self.using_proba
+
+        aggregation_function = self.aggregation_function
+
+        # Save aggregation_function if not None & level_save > LOW
+        if (self.aggregation_function is not None) and (self.level_save in ['MEDIUM', 'HIGH']):
+            # Manage paths
+            aggregation_function_path = os.path.join(self.model_dir, "aggregation_function.pkl")
+            # Save as pickle
+            with open(aggregation_function_path, 'wb') as f:
+                pickle.dump(self.aggregation_function, f)
```

Comment:
How does this work if the aggregation function definition change ? We have the same problem for preprocessing functions.